dxr/vcs.py

```diff
@@ -0,0 +1,322 @@
+
+    def generate_diff(self, path):
+        info = self.have[path]
+        haveRev = info['haveRev']
+        prevRev = str(int(haveRev) - 1)
+        return (self.upstream + info['depotFile'] + '?ac=19&rev1=' + prevRev +
+                '&rev2=' + haveRev)
+
+    def generate_blame(self, path):
+        info = self.have[path]
+        return self.upstream + info['depotFile'] + '?ac=193'
+
+    def generate_log(self, path):
+        info = self.have[path]
+        return self.upstream + info['depotFile'] + '?ac=22#' + info['haveRev']
+
+    def display_rev(self, path):
+        info = self.have[path]
+        return '#' + info['haveRev']
+
+every_vcs = [Mercurial, Git, Perforce]
+
+def tree_to_repos(tree):
+    """Given a TreeConfig, return a mapping {root: Vcs object} where root is a
+    directory under tree.source_folder where root is a directory under
+    tree.source_folder. Traversal of the returned mapping follows the order of
+    deepest directory first.
+
+    :arg tree: TreeConfig object representing a source code tree
+
+    """
+    sources = {}
+    # Find all of the VCSs in the source directory:
+    # We may see multiple VCS if we use git submodules, for example.
+    for cwd, dirs, files in os.walk(tree.source_folder):
+        for vcs in every_vcs:
+            attempt = vcs.claim_vcs_source(cwd, dirs, tree)
+            if attempt is not None:
+                sources[attempt.root] = attempt
+
+    # It's possible that the root of the tree is not a VCS by itself, so walk up
+    # the hierarchy until we find a parent folder that is a VCS. If we can't
+    # find any, then no VCSs exist for the top level of this repository.
+    directory = tree.source_folder
+    while directory != '/' and directory not in sources:
+        directory = os.path.dirname(directory)
+        for vcs in every_vcs:
+            attempt = vcs.claim_vcs_source(directory, os.listdir(directory), tree)
+            if attempt is not None:
+                sources[directory] = attempt
+    lookup_order = sorted(sources.keys(), key=len, reverse=True)
+    # We want to make sure that we look up source repositories by deepest
+    # directory first.
+    ordered_sources = OrderedDict()
+    for key in lookup_order:
+        ordered_sources[key] = sources[key]
+    return ordered_sources
+
+
+class VcsTree(object):
+    """This class offers a way to obtain Vcs objects for any file within a
+    given tree."""
+    def __init__(self, tree):
+        self.tree = tree
+        self.repos = tree_to_repos(tree)
+        self._path_cache = {}
+
+    def vcs_for_path(self, path):
+        """Given a tree and a path in the tree, find a source repository we
+        know about that claims to track that file.
+
+        :arg tree: TreeConfig object representing a source code tree
+        :arg string path: a path to a file (not a folder)
+        """
+        if path in self._path_cache:
+            return self._path_cache
```

Comment:
Yes, this is a bad mistake. I'll add in a duplicated query to one of the /rev/ tests, which will test that the cache is not broken.